settings: Add option to customize channel notifications in Personal settings > Notifications.

saubhagya-patel opened the pull request:
This PR adds an option to change the notification settings of a channel from default to custom in the Notifications menu of Personal settings.

<!-- Describe your pull request here.-->

**Placeholder behavior of the widget**: If no channel has been added to the notification table (there are no channels for which the user has customized the notifications), the initial placeholder value is "Customize a channel". Once a channel is added to the table, the placeholder is changed to "Customize another channel". - [CZO thread](https://chat.zulip.org/#narrow/channel/101-design/topic/notification.20settings.20for.20muted.20streams/near/2051962).


Fixes: #19849.
[CZO thread](https://chat.zulip.org/#narrow/stream/101-design/topic/notification.20settings.20for.20muted.20streams/near/1256417).


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Added option to customize channel notifications in Personal settings > Notifications.</summary>

**Light mode:**

- When no channel is added to the Notification table.
![image](https://github.com/user-attachments/assets/2e50c4a8-b66d-413b-bca9-e597e3ccde01)

- When a channel is added to the Notification table.
![image](https://github.com/user-attachments/assets/ee2f5587-b634-4807-97b2-36c95becbcc4)


**Dark mode:**

- When no channel is added to the Notification table.
![image](https://github.com/user-attachments/assets/47ff916f-5080-461d-b36e-f176aea2f7a0)

- When a channel is added to the Notification table.
![image](https://github.com/user-attachments/assets/b0d60f69-f1f1-4128-9944-4f40754827b4)


</details>

<details>
<summary>Behavior of adding a channel to customize notifications.</summary>

**Light mode:**
![light](https://github.com/user-attachments/assets/c59e9da8-b6ad-46ff-9658-c4185e79dac9)

**Dark mode:**
![dark](https://github.com/user-attachments/assets/b5df7fb5-51d4-40cc-a0be-eaf46668d3d8)

</details>


---

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +179,77 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+                name: sub.name,
+                unique_id: stream.stream_id,
+                stream: sub,
+            };
```
Why not just have this be a list of `StreamSubscription` objects? I don't see how the copy is helpful.

saubhagya-patel replied:
Please, correct me if I am wrong, but, I think we can't have a list just of `StreamSubscription` objects as the `get_options` function callback expects a function whose return value is an array of  `Option` objects ( [file reference](https://github.com/zulip/zulip/blob/main/web/src/dropdown_widget.ts#L42) ): 
```ts
Option = {
    unique_id: number | string;
    name: string;
    is_setting_disabled?: boolean;
    stream?: StreamSubscription;
};
```
Hence, we need to have the fields `unique_id` and `name` also in the return value of the function `get_streams_to_customize_notifications()` in addition to the `StreamSubscription` object.

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +179,77 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+        .map(function () {
+            return Number.parseInt($(this).attr("data-stream-id")!, 10);
+        })
+        .toArray();
```
We should use a set data structure, not an array, to avoid quadratic work.

saubhagya-patel replied:
That sounds good. I have used a Set data structure now.
Moreover, I have moved the calculation of `stream_ids_with_custom_notifications` to the `rerender_ui()` function. So, we don't need to find the value of `stream_ids_with_custom_notifications` every time dropdown is opened.

alya commented:
@sahil839 Adding this to your post-release review queue. Thanks!

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +182,66 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+
+    $("#customizable_stream_notifications_table")
+        .find("input[type='checkbox']")
+        .prop("disabled", false);
```
Shouldn't some of these values potentially be disabled if they would be had that channel already been customized? E.g. if push notifications are disabled for the server, I'd expect that column to still be disabled here.

saubhagya-patel replied:
A check for push notifications is added in the following push, so the checkbox should be disabled as per the setting.
Channels with already customized notifications are filtered out and not displayed in the options, hence, I believe we don't need any other check.
Please let me know if we need any. Thank you.

sahil839 commented on web/styles/settings.css:
```diff
@@ -495,6 +495,10 @@ input[type="checkbox"] {
 }
 
+#customizable_stream_notifications_table .dropdown-widget-button {
+    width: 225px;
```
This should be set in `em` units so that width increases with font size.

saubhagya-patel replied:
Ooh my bad, changed in the following push, will remember this in the future. Thank you!

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +182,72 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+    assert(customize_stream_notifications_widget !== undefined);
+    const stream_id = customize_stream_notifications_widget.value();
+    assert(typeof stream_id === "number");
+    stream_notification_setting_changed(target, stream_id);
```
Is this function needed? This code can directly be added in the caller.

saubhagya-patel replied:
Sure, that sounds right.

sahil839 commented on web/src/settings.ts:
```diff
@@ -114,6 +114,8 @@ export function build_page(): void {
         twenty_four_hour_time_values: settings_config.twenty_four_hour_time_values,
         general_settings: settings_config.all_notifications(user_settings).general_settings,
         notification_settings: settings_config.all_notifications(user_settings).settings,
+        customize_stream_notifications_options:
```
I think `stream_specific_notification_settings` is a better variable name since the notification settings are same, we are just providing a widget here to customize them for a stream.

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +182,72 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+    const user_id = people.my_current_user_id();
+
+    return stream_data
+        .get_streams_for_user(user_id)
```
This includes archived channels as well. I don't think we want to show them here.

sahil839 commented on web/src/settings_notifications.ts:
```diff
@@ -171,6 +182,72 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+        get_options: get_streams_to_customize_notifications,
+        item_click_callback: change_state_of_customize_stream_notifications_widget,
+        $events_container: $("#user-notification-settings .notification-settings-form"),
+        unique_id_type: dropdown_widget.DataTypes.NUMBER,
```
This gives error in development server. This should be changed to "number".

saubhagya-patel replied:
This was caused by the commit [dropdown_widget: Convert TypeScript enum to union.](https://github.com/zulip/zulip/commit/dd71087b5f8f30f8ca06d42522ba7e1ad5a91d61).
I didn't update this PR after the commit. However, this is fixed in the following push. Thank you!

sahil839 commented on web/templates/settings/notification_settings.hbs:
```diff
@@ -49,6 +49,21 @@
+                          setting_name=this
+                          is_checked=false
+                          is_disabled=true
+                          is_mobile_checkbox=false
```
This should be true for mobile checkbox so that the tooltip is visible if it is disabled.

saubhagya-patel replied:
Sure, I have changed the approach that was used here. Please take a look.

sahil839 commented on web/src/settings_config.ts:
```diff
@@ -990,6 +990,22 @@ export function get_notifications_table_row_data(
+            is_mobile_checkbox: false,
+        };
+        if (setting_name === "push_notifications") {
+            checkbox.is_disabled = !realm.realm_push_notifications_enabled;
```
The checkbox for mobile push notifications is enabled by default when no stream is selected if push notifications are allowed by the server.

saubhagya-patel replied:
This is fixed in the following push. Thank you.

sahil839 commented on web/src/settings.ts:
```diff
@@ -115,6 +115,8 @@ export function build_page(): void {
         twenty_four_hour_time_values: settings_config.twenty_four_hour_time_values,
         general_settings: settings_config.all_notifications(user_settings).general_settings,
         notification_settings: settings_config.all_notifications(user_settings).settings,
+        stream_specific_notification_settings:
```
Now that the object with details of checkbox is passed, and not only the setting names, we should rename this variable. Something like `custom_stream_specific_notification_settings`.

sahil839 reviewed (commented):
Posted couple of comments. Looks good otherwise.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2028539241) comment, the [`settings_config.get_custom_stream_specific_notifications_table_row_data()`](https://github.com/zulip/zulip/pull/33181/files#diff-30259fce41ce056dbd4ea251d7849d395d45438a1b696287c31ff6191388459dR993-R1010) function is updated to return disabled checkboxes by default for all stream-specific notification settings. This prevents the condition stated [here](https://github.com/zulip/zulip/pull/33181#discussion_r2028539241).
However, please note that for mobile push notifications, the `is_mobile_checkbox` value is set based on the `realm_push_notifications_enabled` setting to handle the mobile push notifications disabled tooltip display correctly:
  - If `realm_push_notifications_enabled` is `false`, all checkboxes are disabled initially, and the mobile push notifications disabled tooltip is shown (as expected when the checkbox is disabled and marked as an `is_mobile_checkbox` - [template reference](https://github.com/zulip/zulip/blob/main/web/templates/settings/notification_settings_checkboxes.hbs#L2)).

  - If `realm_push_notifications_enabled` is `true`, all checkboxes are still disabled initially, but the mobile push notifications disabled tooltip should NOT be shown, so the function avoids setting it as an `is_mobile_checkbox`.

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2028542650) comment, `stream_specific_notification_settings` is renamed to `custom_stream_specific_notification_settings`.

@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

sahil839 commented on web/src/settings_config.ts:
```diff
@@ -990,6 +990,24 @@ export function get_notifications_table_row_data(
+            // Prevents the mobile push notifications disabled tooltip from showing
+            // unnecessarily when the checkbox is initially disabled due to no stream
+            // being selected, even though push notifications are enabled.
+            checkbox.is_mobile_checkbox = !realm.realm_push_notifications_enabled;
```
This works but maybe we can do better than passing `is_mobile_checkbox` as `false` for push notification. We can change the variable used to show the tooltip to be something like `push_notifications_disabled` and then tooltip can be showed only if both `is_mobile_checkbox` and `push_notifications_disabled` are true.

saubhagya-patel replied:
@sahil839, please correct me if I am wrong, but I am a little confused here. Do you mean to say that we should pass an extra variable (`push_notifications_disabled`) in the `notification_settings_checkboxes.hbs` file to address this and keep the variable `is_disabled` too?
Currently, the tooltip is shown if both `is_mobile_checkbox` and `is_disabled` are true. However, we use the `is_disabled` variable for `push_notifications` only (in the `settings_config.get_notifications_table_row_data()` - ([ref](https://github.com/zulip/zulip/blob/main/web/src/settings_config.ts#L959))), so, if we change `is_disabled` -> `push_notifications_disabled` it will result in same behavior.

Could you please give a little more context around this?
Thank you.

sahil839 replied:
As you said, we use `is_disabled` only for push notifications, we should rename that to `push_notifications_disabled` and it would be set to true or false depending upon the value of `realm.realm_push_notifications_enabled`. And `is_mobile_checkbox` value should not depend on `realm.realm_push_notifications_enabled`, it should be true for push notifications checkbox.

Then these two variables can be used to compute whether to show the tooltip or not.

saubhagya-patel replied:
From what I understood from the above comment, I used the above approach to rename `is_disabled` -> `push_notifications_disabled` and set its value based upon the value of `realm.realm_push_notifications_enabled`. The value of `is_mobile_checkbox` was set to true for the push notifications checkbox. The function `get_custom_stream_specific_notifications_table_row_data()` was updated as follows:

<details>
<summary>Updated <code>get_custom_stream_specific_notifications_table_row_data()</code> function.  </summary>

- Here, I am only including changes to the function; the changes to other files, like `notification_settings_checkboxes.hbs`, for instance, are not included here.

```diff
diff --git a/web/src/settings_config.ts b/web/src/settings_config.ts
index b6ffdcf675..d5965ecb4f 100644
--- a/web/src/settings_config.ts
+++ b/web/src/settings_config.ts
@@ -951,7 +951,7 @@ type Settings = UserSettings | RealmDefaultSettings;
 type PageParamsItem = keyof Settings;
 type NotificationSettingCheckbox = {
     setting_name: string;
-    is_disabled: boolean;
+    push_notifications_disabled: boolean;
     is_checked: boolean;
     is_mobile_checkbox: boolean;
 };
@@ -994,12 +994,13 @@ export function get_custom_stream_specific_notifications_table_row_data(): Notif
     return stream_specific_notification_settings.map((setting_name) => {
         const checkbox = {
             setting_name,
-            is_disabled: true,
+            push_notifications_disabled: true,
             is_checked: false,
             is_mobile_checkbox: false,
         };
         if (setting_name === "push_notifications") {
-            checkbox.is_mobile_checkbox = !realm.realm_push_notifications_enabled;
+            checkbox.is_mobile_checkbox = true;
+            checkbox.push_notifications_disabled = !realm.realm_push_notifications_enabled;
         }
         return checkbox;
     });
```

</details>


However, I found a bug with this approach.
If push notifications are enabled for a realm, the checkbox for mobile push notifications is enabled by default when no stream is selected, similar to [this](https://github.com/zulip/zulip/pull/33181#discussion_r2028539241) comment.

We cannot always pass `push_notifications_disabled` as true, since if push notifications are enabled for a realm, it will show the mobile push notifications disabled tooltip when hovering on the mobile push notifications checkbox.

<details>
<summary> Screenshots. </summary>

- Bug with the approach - If push notifications are enabled for a realm, the checkbox for mobile push notifications is enabled by default when no stream is selected, similar to [this](https://github.com/zulip/zulip/pull/33181#discussion_r2028539241) comment.
![image](https://github.com/user-attachments/assets/1c97c499-8d72-4e66-9b94-d73ce3544810)


- If we always pass `push_notifications_disabled` as true, it will show the mobile push notifications disabled tooltip when hovering on the mobile push notifications checkbox, even though push notifications are enabled for the realm.
![image](https://github.com/user-attachments/assets/6676061c-9ec3-4c71-b433-8f23c05270e7)

</details>

Please let me know if I missed something or if something was inappropriate.
Thank you.

sahil839 replied:
I guess we can keep both `is_disabled` and `push_notifications_disabled` variables. `is_disabled` can be used to disable the checkbox and `push_notifications_disabled` to show the tooltip.

saubhagya-patel replied:
Yeah, that sounds good. I have implemented this in the following push.
Thank you.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2034619854) and the following comments, to have a variable `push_notifications_disabled` for controlling the mobile push notifications disabled tooltip, I have made the required changes in a prep commit: [settings: Add `push_notifications_disabled` in notification checkboxes.](https://github.com/zulip/zulip/pull/33181/commits/ae5d48797608bea02c83f617c081d76f30aa12d3).


@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are needed.
Thank you.

sahil839 commented:
Looks good to me. @alya do you want to have a look at this or should I tag this ready for review by Tim?

alya commented:
I think when a long channel name is selected, we'd ideally have the same alignment in the dropdown (not centered).

![Screenshot 2025-04-29 at 00 56 28](https://github.com/user-attachments/assets/6e2573b4-4b51-409d-beca-b4cafde87bbf)

I also discovered [#issues > long one-word channel name breaks announcement settings](https://chat.zulip.org/#narrow/channel/9-issues/topic/long.20one-word.20channel.20name.20breaks.20announcement.20settings/with/2163246) while playing with this.

alya commented:
I think we don't want a focus ring when a user clicks on the dropdown -- it should only be shown if you're using the keyboard.

alya commented:
Looks good to me otherwise! (I haven't tested that the settings actually work as expected.)

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/33181#issuecomment-2837858583) comment regarding the focus ring around the dropdown, the CSS is updated to show focus outline when using keyboard and not when a user clicks on the dropdown.
  - <details>
    <summary>Screenshots of the focus ring behavior.</summary>

    - No focus on using the mouse.  
      ![mouse no focus](https://github.com/user-attachments/assets/b64686fc-6a1a-4fff-890b-8ea96ee8f337)

    - Focus when using the keyboard.  
      ![keyboard focus](https://github.com/user-attachments/assets/f65adf32-2789-4ccf-9c8c-3908024efe99)
    </details>




- Moreover, could you provide a little more context around [this](https://github.com/zulip/zulip/pull/33181#issuecomment-2837856995) comment, I am not completely sure about the expected behavior.

  > I think when a long channel name is selected, we'd ideally have the same alignment in the dropdown (not centered).
  > 
  > ![Screenshot 2025-04-29 at 00 56 28](https://private-user-images.githubusercontent.com/2090066/438631249-6e2573b4-4b51-409d-beca-b4cafde87bbf.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDcxMzM1OTQsIm5iZiI6MTc0NzEzMzI5NCwicGF0aCI6Ii8yMDkwMDY2LzQzODYzMTI0OS02ZTI1NzNiNC00YjUxLTQwOWQtYmVjYS1iNGNhZmRlODdiYmYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDUxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA1MTNUMTA0ODE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MzZlYTdkMTViZmVjNmRhNDkyMWM5ZWU3YjEzMDVhZGYyMDEyMGRmNTFmMTYxNDVkMDU3ODBhMGE0ZWY1MTIzYiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ._sxVEGbXMPum8cuMYhRcn7Ilfzr9QUB3slrqDk-JRp8)
  > 
  > I also discovered [#issues > long one-word channel name breaks announcement settings](https://chat.zulip.org/#narrow/channel/9-issues/topic/long.20one-word.20channel.20name.20breaks.20announcement.20settings/with/2163246) while playing with this.
  
  Do we need to have the same format as introduced in #34517?
  Something like the following:
  ![image](https://github.com/user-attachments/assets/d04b1d36-1e3c-486b-b541-72409d9556ff)

  

@alya, I have made the required changes. I hope the updated changes look good to you.
I have a question regarding [this](https://github.com/zulip/zulip/pull/33181#issuecomment-2837856995) comment, which I have mentioned above. Once I understand it, I'll update the PR accordingly.
Thank you.

saubhagya-patel commented:
Upon further thinking, can we make a separate class to apply the styles to the dropdown-widget for long channel names, as in #34517, to avoid repetition, maybe something like `truncate-long-name-in-dropdown`? I am not sure about the classname, though.

Then we can pass it as a `custom_class` in the `dropdown_widget_with_label` partial, a little similar to what we did for `push_notifications_disabled`.
@sahil839, could you please share your thoughts on this.

alya commented:
Yep, those visuals for long channel names look good to me -- thanks!

sahil839 commented on web/styles/settings.css:
```diff
@@ -847,6 +873,7 @@ input[type="checkbox"] {
         .decorated-channel-name {
             overflow-x: hidden;
             text-overflow: ellipsis;
+            white-space: nowrap;
```
This change should be in a prep commit.

sahil839 commented on web/styles/settings.css:
```diff
@@ -489,6 +489,32 @@ input[type="checkbox"] {
+        .decorated-channel-name {
+            overflow-x: hidden;
+            text-overflow: ellipsis;
+            white-space: nowrap;
```
As, you said we can probably create a shared class and use that CSS for all dropdown widgets with streams.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2090624947) comment, the style `white-space: nowrap;` is added in a prep commit: [css: Set white-space: nowrap for channel dropdowns in org_settings.](https://github.com/zulip/zulip/pull/33181/commits/0771bb363bc8a4c3348caad15f693a9e7b93a812).

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2090631798) comment, a new class `abbreviate-dropdown-value` is added in a separate commit as well: [css: Add class abbreviate-dropdown-value.](https://github.com/zulip/zulip/pull/33181/commits/53ad45a8c5186876ce0f156897007acedb50c176).
  
  <details>
  <summary> <code>git grep</code> of introduced class. </summary>
  
  - Before:
  
  ```bash
  $ git grep abbreviate-dropdown-value
  
  ```
  
  - After:
  
  ```bash
  $ git grep abbreviate-dropdown-value
  web/styles/settings.css:.abbreviate-dropdown-value {
  web/templates/settings/notification_settings.hbs:                    <td class="abbreviate-dropdown-value">
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="abbreviate-dropdown-value"}}
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="abbreviate-dropdown-value"}}
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="abbreviate-dropdown-value"}}
  ```
  </details>
  
@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

sahil839 commented on web/styles/settings.css:
```diff
@@ -415,6 +415,26 @@ input[type="checkbox"] {
     }
 }
 
+.abbreviate-dropdown-value {
```
Can this class be named such that we know this is being used for stream dropdowns? I think that would be better since we anyways have the CSS to abbreviate names on `decorated-channel-name`.

saubhagya-patel replied:
Can we use something like `abbreviate-stream-dropdown-value`?

sahil839 replied:
I think there is no need to mention `abbreviate`, we can just have something like `decorated-stream-name-dropdown-widget`.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following change was made:

- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2103878518) comment, the classname `abbreviate-dropdown-value` is updated to `decorated-stream-name-dropdown-widget`.
  
  <details>
  <summary> <code>git grep</code> of introduced class. </summary>
  <br />
  
  - Before:
  
  ```bash
  $ git grep decorated-stream-name-dropdown-widget
  
  ```
  
  - After:
  
  ```bash
  $ git grep decorated-stream-name-dropdown-widget
  web/styles/settings.css:.decorated-stream-name-dropdown-widget {
  web/templates/settings/notification_settings.hbs:                    <td class="decorated-stream-name-dropdown-widget">
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="decorated-stream-name-dropdown-widget"}}
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="decorated-stream-name-dropdown-widget"}}
  web/templates/settings/organization_settings_admin.hbs:                  custom_classes="decorated-stream-name-dropdown-widget"}}
  ```
  </details>
  
@sahil839, I have made the required changes. I hope the updated changes look good to you. Please let me know if any further adjustments are required.
Thank you.

timabbott commented on web/templates/settings/stream_specific_notification_row.hbs:
```diff
@@ -16,6 +16,7 @@
           is_checked=(lookup ../stream this)
           is_disabled=(lookup ../is_disabled this)
           is_mobile_checkbox=(eq this "push_notifications")
+          push_notifications_disabled=../push_notifications_disabled
```
Does this syntax work? This is not at all how `is_disabled` is accesed.

saubhagya-patel replied:
Yes, this syntax works based on how we pass `is_disabled` in [`web/src/settings_notifications.ts`](https://github.com/zulip/zulip/blob/main/web/src/settings_notifications.ts#L74).

timabbott commented on web/styles/settings.css:
```diff
@@ -834,24 +854,6 @@ input[type="checkbox"] {
-            overflow-x: hidden;
-            text-overflow: ellipsis;
-            white-space: nowrap;
-        }
```
This change makes this CSS no longer apply to the `realm_default_code_block_language`  dropdown widget. Is that correct?

saubhagya-patel replied:
Yes, it is correct, I don't think we will need to abbreviate the code block language.

timabbott commented on web/styles/settings.css:
```diff
@@ -834,24 +854,6 @@ input[type="checkbox"] {
 #organization-settings {
     .dropdown-widget-button {
         color: hsl(0deg 0% 33%);
```
This color being individually written the same for multiple things, with no `light-dark` color, seems like a code smell. Should this property just be on `dropdown-widget-button` or this CSS deleted?

timabbott commented on web/src/settings_config.ts:
```diff
@@ -1012,6 +1012,22 @@ export function get_notifications_table_row_data(
+        };
+        if (setting_name === "push_notifications") {
+            checkbox.is_mobile_checkbox = true;
+        }
```
This is probably better done as `is_mobile_checkbox: setting_name === "push_notifications"`.

timabbott commented on web/src/settings_notifications.ts:
```diff
@@ -211,6 +221,56 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+
+    const $customizable_stream_notifications_table = $("#customizable_stream_notifications_table");
+
+    $customizable_stream_notifications_table.find("input[type='checkbox']").prop("disabled", false);
```
I think you can use blank lines better in this area to make it more readable. I would probably be doing a blank line after `stopPropagation`, but not one just before this line.

saubhagya-patel commented on web/src/settings_config.ts:
```diff
@@ -1012,6 +1012,22 @@ export function get_notifications_table_row_data(
 }
 
+export function get_custom_stream_specific_notifications_table_row_data(): NotificationSettingCheckbox[] {
+    return stream_specific_notification_settings.map((setting_name) => {
```
Updated with the following comment:

> Returns an array of NotificationSettingCheckbox for the special row that
> allows adding new configuration for a previously uncustomized channel.

saubhagya-patel commented on web/src/settings_notifications.ts:
```diff
@@ -211,6 +221,56 @@ function stream_notification_setting_changed(target: HTMLInputElement): void {
+        if (!stream_ids_with_custom_notifications.has(stream.stream_id)) {
+            return true;
+        }
+        return false;
```
Yes, refactored it in the following push.

saubhagya-patel commented:
Following the review received, I have updated this PR. The following changes were made:

- Refactored code following the comments ([[1]](https://github.com/zulip/zulip/pull/33181#discussion_r2131063850), [[2]](https://github.com/zulip/zulip/pull/33181#discussion_r2131068701), [[3]](https://github.com/zulip/zulip/pull/33181#discussion_r2131077971)).
- Following [this](https://github.com/zulip/zulip/pull/33181#discussion_r2131065588) comment, a comment is added to explain the function `get_custom_stream_specific_notifications_table_row_data()`.

- Here, is the diff for the updations:

```diff
$ git diff origin/add-stream-custom-notifications 
diff --git a/web/src/settings_config.ts b/web/src/settings_config.ts
index 4fd03a504b..b974857dee 100644
--- a/web/src/settings_config.ts
+++ b/web/src/settings_config.ts
@@ -1013,17 +1013,16 @@ export function get_notifications_table_row_data(
 }
 
 export function get_custom_stream_specific_notifications_table_row_data(): NotificationSettingCheckbox[] {
+    // Returns an array of NotificationSettingCheckbox for the special row that
+    // allows adding new configuration for a previously uncustomized channel.
     return stream_specific_notification_settings.map((setting_name) => {
         const checkbox = {
             setting_name,
             is_disabled: true,
             is_checked: false,
-            is_mobile_checkbox: false,
+            is_mobile_checkbox: setting_name === "push_notifications",
             push_notifications_disabled: !realm.realm_push_notifications_enabled,
         };
-        if (setting_name === "push_notifications") {
-            checkbox.is_mobile_checkbox = true;
-        }
         return checkbox;
     });
 }
diff --git a/web/src/settings_notifications.ts b/web/src/settings_notifications.ts
index 7af327be01..b1f92db8c9 100644
--- a/web/src/settings_notifications.ts
+++ b/web/src/settings_notifications.ts
@@ -228,13 +228,13 @@ function change_state_of_customize_stream_notifications_widget(
     dropdown.hide();
     event.preventDefault();
     event.stopPropagation();
+
     assert(customize_stream_notifications_widget !== undefined);
     customize_stream_notifications_widget.render();
     const stream_id = customize_stream_notifications_widget.value();
     assert(typeof stream_id === "number");
 
     const $customizable_stream_notifications_table = $("#customizable_stream_notifications_table");
-
     $customizable_stream_notifications_table.find("input[type='checkbox']").prop("disabled", false);
 
     if (!realm.realm_push_notifications_enabled) {
@@ -252,12 +252,9 @@ function change_state_of_customize_stream_notifications_widget(
 }
 
 function get_streams_to_customize_notifications(): dropdown_widget.Option[] {
-    return stream_data.get_options_for_dropdown_widget().filter(({stream}) => {
-        if (!stream_ids_with_custom_notifications.has(stream.stream_id)) {
-            return true;
-        }
-        return false;
-    });
+    return stream_data
+        .get_options_for_dropdown_widget()
+        .filter(({stream}) => !stream_ids_with_custom_notifications.has(stream.stream_id));
 }
 
 function render_customize_stream_notifications_widget(): void {
```


Thank you.

timabbott commented:
Merged, thanks @saubhagya-patel and @sahil839!
